app/src/main/cpp/skyline/gpu/memory_manager.cpp

```diff
@@ -0,0 +1,126 @@
+            .vkFreeMemory = deviceDispatcher->vkFreeMemory,
+            .vkMapMemory = deviceDispatcher->vkMapMemory,
+            .vkUnmapMemory = deviceDispatcher->vkUnmapMemory,
+            .vkFlushMappedMemoryRanges = deviceDispatcher->vkFlushMappedMemoryRanges,
+            .vkInvalidateMappedMemoryRanges = deviceDispatcher->vkInvalidateMappedMemoryRanges,
+            .vkBindBufferMemory = deviceDispatcher->vkBindBufferMemory,
+            .vkBindImageMemory = deviceDispatcher->vkBindImageMemory,
+            .vkGetBufferMemoryRequirements = deviceDispatcher->vkGetBufferMemoryRequirements,
+            .vkGetImageMemoryRequirements = deviceDispatcher->vkGetImageMemoryRequirements,
+            .vkCreateBuffer = deviceDispatcher->vkCreateBuffer,
+            .vkDestroyBuffer = deviceDispatcher->vkDestroyBuffer,
+            .vkCreateImage = deviceDispatcher->vkCreateImage,
+            .vkDestroyImage = deviceDispatcher->vkDestroyImage,
+            .vkCmdCopyBuffer = deviceDispatcher->vkCmdCopyBuffer,
+            .vkGetBufferMemoryRequirements2KHR = deviceDispatcher->vkGetBufferMemoryRequirements2,
+            .vkGetImageMemoryRequirements2KHR = deviceDispatcher->vkGetImageMemoryRequirements2,
+            .vkBindBufferMemory2KHR = deviceDispatcher->vkBindBufferMemory2,
+            .vkBindImageMemory2KHR = deviceDispatcher->vkBindImageMemory2,
+            .vkGetPhysicalDeviceMemoryProperties2KHR = instanceDispatcher->vkGetPhysicalDeviceMemoryProperties2,
+        };
+        VmaAllocatorCreateInfo allocatorCreateInfo{
+            .physicalDevice = *gpu.vkPhysicalDevice,
+            .device = *gpu.vkDevice,
+            .instance = *gpu.vkInstance,
+            .pVulkanFunctions = &vulkanFunctions,
+            .vulkanApiVersion = GPU::VkApiVersion,
+        };
+        ThrowOnFail(vmaCreateAllocator(&allocatorCreateInfo, &vmaAllocator));
+        // TODO: Use VK_KHR_dedicated_allocation when available (Should be on Adreno GPUs)
+    }
+
+    MemoryManager::~MemoryManager() {
+        vmaDestroyAllocator(vmaAllocator);
+    }
+
+    std::shared_ptr<StagingBuffer> MemoryManager::AllocateStagingBuffer(vk::DeviceSize size) {
+        vk::BufferCreateInfo bufferCreateInfo{
+            .size = size,
+            .usage = vk::BufferUsageFlagBits::eTransferSrc,
+            .sharingMode = vk::SharingMode::eExclusive,
+            .queueFamilyIndexCount = 1,
+            .pQueueFamilyIndices = &gpu.vkQueueFamilyIndex,
+        };
+        VmaAllocationCreateInfo allocationCreateInfo{
+            .flags = VMA_ALLOCATION_CREATE_MAPPED_BIT,
+            .usage = VMA_MEMORY_USAGE_CPU_ONLY,
+        };
+
+        VkBuffer buffer;
+        VmaAllocation allocation;
+        VmaAllocationInfo allocationInfo;
+        ThrowOnFail(vmaCreateBuffer(vmaAllocator, &static_cast<const VkBufferCreateInfo &>(bufferCreateInfo), &allocationCreateInfo, &buffer, &allocation, &allocationInfo));
```

Comment:
Sadly, it cannot as the cast from `vk::ImageCreateInfo` (Vulkan-Hpp type) to `VkImageCreateInfo` (Vulkan C type) is done by defining `vk::ImageCreateInfo::operator VkImageCreateInfo &()` which needs to be casted to using an explicit cast prior to taking a pointer value since the pointer value becomes `vk::ImageCreateInfo*` that cannot be implicitly casted to `VkImageCreateInfo*` as pointer values do not inherit the implicit casts of their underlying types.